### ISSUE TITLE
clean old or uninstalled AUR packages in ~/.cache/pikaur/pkg

### ISSUE THREAD

**Issue by h3fang** [actionless/pikaur#165]
Can your make pikaur clean old or uninstalled AUR packages in ~/.cache/pikaur/pkg with `pikaur -Sc` ?

**Comment by hfenigma** [comment]
`pikaur -Scc` removes the entire pkg folder
`pikaur -Sc` seems to do nothing, leaving old and uninstalled packages there

**Comment by hfenigma** [comment]
Actually I wanted something similar to `pacman -Sc`, which removes old and uninstalled packages, leaving the latest installed there, so if something goes wrong, I can easily downgrade it.

**Comment by AladW** [comment]
Pretty hard to guess from the little/vague information what's meant, but I suppose:
* Set `PKGDEST` to a `CacheDir` like `/var/cache/pacman/<some directory>`
  * Can be tidied with `paccache` or `pacman -Sc` (`CleanMethod = KeepInstalled`)
* Use `git clean` for remove old source files
  * Batch for all in `pikaur` cache with `find <pikaur dir> -execdir git clean -xf`
  * Alternatively set `SRCDEST` 

In other words these are things that already can easily done outside of `pikaur`. I suggest adding a note to the readme.

**Comment by actionless** [comment]
@AladW as far as i understood the original report the whole point is about `KeepInstalled` option for AUR packages

**Comment by AladW** [comment]
For what, source directories? Seems pretty pointless when these are few kb in size.

**Comment by actionless** [comment]
for built packages

**Comment by AladW** [comment]
An AUR package is a package like any other, right? If you want to run `paccache` without it affecting your repo packages, you can just use a different directory than `/var/cache/pacman/pkg`. See `paccache -c <path>`.

**Comment by actionless** [comment]
@hfenigma  if  `paccache -c ~/.cache/pikaur/pkg/` does what you expect to?

**Comment by hfenigma** [comment]
For example, I have the following AUR packages installed
> $ pacman -Qm
brightnessctl 0.3.2-1
openblas-lapack 0.2.20-2
pikaur 0.10.3-1
visual-studio-code-bin 1.23.1-1
xss-lock-git 0.3.0.r4.g1e158fb-1
zotero 5.0.47-1

And these are what's in ~/.cache/pikaur/pkg/
>$ ls ~/.cache/pikaur/pkg/
total 48620
-rw-r--r-- 1 enigma enigma    77984 May 15 18:32 aurman-2.9.26-1-any.pkg.tar.xz
-rw-r--r-- 1 enigma enigma    16620 May 14 02:39 brightnessctl-0.3.2-1-x86_64.pkg.tar.xz
-rw-r--r-- 1 enigma enigma  3111184 May 13 21:36 openblas-lapack-0.2.20-2-any.pkg.tar.xz
-rw-r--r-- 1 enigma enigma   152304 May 12 10:31 pikaur-0.10.2-1-any.pkg.tar.xz
-rw-r--r-- 1 enigma enigma   152372 May 15 11:45 pikaur-0.10.3-1-any.pkg.tar.xz
-rw-r--r-- 1 enigma enigma 46257196 May 13 19:33 visual-studio-code-bin-1.23.1-1-x86_64.pkg.tar.xz

I want `some command` to remove `aurman-2.9.26-1-any.pkg.tar.xz` and `pikaur-0.10.2-1-any.pkg.tar.xz` and leave all the others alone

As you suggested, I tried `paccache -c ~/.cache/pikaur/pkg/`
> $ paccache -c ~/.cache/pikaur/pkg/
==> ERROR: no operation specified (use -h for help)

Then I looked into the help
>$ paccache --help
paccache (pacman) v5.0.2
A flexible pacman cache cleaning utility.
Usage: paccache <operation> [options] [targets...]
  Operations:
    -d, --dryrun          perform a dry run, only finding candidate packages.
    -m, --move <dir>      move candidate packages to "dir".
    -r, --remove          remove candidate packages.
  Options:
    -a, --arch <arch>     scan for "arch" (default: all architectures).
    -c, --cachedir <dir>  scan "dir" for packages. can be used more than once.
                          (default: read from /etc/pacman.conf).
    -f, --force           apply force to mv(1) and rm(1) operations.
    -h, --help            display this help message and exit.
    -i, --ignore <pkgs>   ignore "pkgs", comma-separated. Alternatively, specify
                          "-" to read package names from stdin, newline-
                          delimited.
    -k, --keep <num>      keep "num" of each package in the cache (default: 3).
        --nocolor         remove color from output.
    -q, --quiet           minimize output
    -u, --uninstalled     target uninstalled packages.
    -v, --verbose         increase verbosity. specify up to 3 times.
    -z, --null            use null delimiters for candidate names (only with -v
                          and -vv).

Then I tried `paccache -rc ~/.cache/pikaur/pkg/`
>$ paccache -rc ~/.cache/pikaur/pkg/
==> no candidate packages found for pruning

I don't know what to do next, my goal is not achieved.
> $ ls ~/.cache/pikaur/pkg/
total 48620
-rw-r--r-- 1 enigma enigma    77984 May 15 18:32 aurman-2.9.26-1-any.pkg.tar.xz
-rw-r--r-- 1 enigma enigma    16620 May 14 02:39 brightnessctl-0.3.2-1-x86_64.pkg.tar.xz
-rw-r--r-- 1 enigma enigma  3111184 May 13 21:36 openblas-lapack-0.2.20-2-any.pkg.tar.xz
-rw-r--r-- 1 enigma enigma   152304 May 12 10:31 pikaur-0.10.2-1-any.pkg.tar.xz
-rw-r--r-- 1 enigma enigma   152372 May 15 11:45 pikaur-0.10.3-1-any.pkg.tar.xz
-rw-r--r-- 1 enigma enigma 46257196 May 13 19:33 visual-studio-code-bin-1.23.1-1-x86_64.pkg.tar.xz

**Comment by hfenigma** [comment]
not working
>$ paccache -rcud ~/.cache/pikaur/pkg/
==> ERROR: cachedir 'ud' does not exist or is not a directory

**Comment by actionless** [comment]
oops, `paccache -rudc ~/.cache/pikaur/pkg/`

**Comment by hfenigma** [comment]
>$ paccache -rudc ~/.cache/pikaur/pkg/
==> ERROR: only one operation may be used at a time

**Comment by actionless** [comment]
@AladW i think we need you here :-)

**Comment by Morganamilo** [comment]
`-r` is remove `-d` is dryrun, you can only do one operation at a time.

**Comment by hfenigma** [comment]
@Morganamilo yes, I know that, see the updated post

**Comment by Morganamilo** [comment]
So whats in that directory? And is everything in there installed?

**Comment by hfenigma** [comment]
nothing changed 
>$ ls ~/.cache/pikaur/pkg/
total 48620
-rw-r--r-- 1 enigma enigma    77984 May 15 18:32 aurman-2.9.26-1-any.pkg.tar.xz
-rw-r--r-- 1 enigma enigma    16620 May 14 02:39 brightnessctl-0.3.2-1-x86_64.pkg.tar.xz
-rw-r--r-- 1 enigma enigma  3111184 May 13 21:36 openblas-lapack-0.2.20-2-any.pkg.tar.xz
-rw-r--r-- 1 enigma enigma   152304 May 12 10:31 pikaur-0.10.2-1-any.pkg.tar.xz
-rw-r--r-- 1 enigma enigma   152372 May 15 11:45 pikaur-0.10.3-1-any.pkg.tar.xz
-rw-r--r-- 1 enigma enigma 46257196 May 13 19:33 visual-studio-code-bin-1.23.1-1-x86_64.pkg.tar.xz

**Comment by Morganamilo** [comment]
Right, thought it might have been changed without you knowing.

You don't have aurman installed at all right? Not via a local or custom repo of some sort?

**Comment by hfenigma** [comment]
not installed, not local (I think this means by makepkg ?), no custom repo
>$ pacman -Qm
brightnessctl 0.3.2-1
openblas-lapack 0.2.20-2
pikaur 0.10.3-1
visual-studio-code-bin 1.23.1-1
xss-lock-git 0.3.0.r4.g1e158fb-1
zotero 5.0.47-1

**Comment by AladW** [comment]
I've always used this approach with my local repo CacheDirs and no issues. In particular I did:
```
paccache -ruk0 -c /var/cache/pacman/custom
```
You can see what logic `paccache` uses here:
https://git.archlinux.org/pacman-contrib.git/tree/src/paccache.sh.in#n297

**Comment by hfenigma** [comment]
Well, it does removed uninstalled packages, but not old ones.
>$ paccache -ruk0 -c ~/.cache/pikaur/pkg
==> finished: 1 packages removed (disk space saved: 76.16 KiB)

>$ ls ~/.cache/pikaur/pkg/
total 48696
-rw-r--r-- 1 enigma enigma    16620 May 14 02:39 brightnessctl-0.3.2-1-x86_64.pkg.tar.xz
-rw-r--r-- 1 enigma enigma  3111184 May 13 21:36 openblas-lapack-0.2.20-2-any.pkg.tar.xz
-rw-r--r-- 1 enigma enigma   152304 May 12 10:31 pikaur-0.10.2-1-any.pkg.tar.xz
-rw-r--r-- 1 enigma enigma   152372 May 15 11:45 pikaur-0.10.3-1-any.pkg.tar.xz
-rw-r--r-- 1 enigma enigma   156484 May 21 12:31 pikaur-0.11-1-any.pkg.tar.xz
-rw-r--r-- 1 enigma enigma 46257196 May 13 19:33 visual-studio-code-bin-1.23.1-1-x86_64.pkg.tar.xz

**Comment by steven-omaha** [comment]
@hfenigma This is the part from my pacman-hook (paccache-clear.hook). For both official and AUR packages, the last 3 packages are kept if the package is still installed, and one package is kept if the package is uninstalled.

`Exec = /usr/bin/env bash -c "/usr/bin/paccache -vrk3; /usr/bin/paccache -vruk1; /usr/bin/paccache --cachedir /var/cache/aur -vrk3; /usr/bin/paccache --cachedir /var/cache/aur -vruk1"`

**Comment by hfenigma** [comment]
@steven-omaha wow, thank you very much

**Comment by actionless** [comment]
@steven-omaha thanks, added to FAQ

**Comment by jjgalvez** [comment]
I know this is old and closed, but as a feature enhancement please consider adding a systemd timer in /usr/lib/systemd/user/ that would trigger paccache with the -c -r option to keep things clean. users could enable it if they thought it was helpful. I just copied the paccache .service and .timer and added modified them to point to the pikaru cache. it would be great if this was part of the standard install

**Comment by KopfKrieg** [comment]
> I know this is old and closed, but as a feature enhancement please consider adding a systemd timer in /usr/lib/systemd/user/ that would trigger paccache with the -c -r option to keep things clean.

What about a pacman hook similar to [pacman-cleanup-hook](https://aur.archlinux.org/packages/pacman-cleanup-hook/)? Maybe calling it pikaur-cleanup-hook ;)

**Comment by actionless** [comment]
@KopfKrieg yeah, could be a thing

**Comment by jjgalvez** [comment]
I didn't know about the hook. Looks like something that would work better on my laptop then the timer. Thanks for pointing this out. I am going to give this a try.




Get Outlook for Android







On Sat, Jun 1, 2019 at 5:37 AM -0400, "Florian" <notifications@github.com> wrote:













I know this is old and closed, but as a feature enhancement please consider adding a systemd timer in /usr/lib/systemd/user/ that would trigger paccache with the -c -r option to keep things clean.



What about a pacman hook similar to pacman-cleanup-hook? Maybe calling it pikaur-cleanup-hook ;)

**Comment by audeoudh** [comment]
> What about a pacman hook similar to [pacman-cleanup-hook](https://aur.archlinux.org/packages/pacman-cleanup-hook/)? Maybe calling it pikaur-cleanup-hook ;)
> 
> Would also solve the problem that timers are running no matter if necessary or not, while a hook only runs when necessary.

You don't want to use paccache with `--uninstalled` option in the pacman hook, and really not combined with `--keep 0`: while debugging an application on your system, you may want to re-install it, and if that is done in two steps (for any reason, e.g. because of an annoying dependency), the uninstalled package is dropped in-between.

I personally fell into the trap once.

So you cannot get rid of old packages you removed long time ago.

**Comment by audeoudh** [comment]
What would be the best default parameters for automatic weekly cleanup?
- Keep 1 version of installed packages, 0 version of uninstalled packages (a bit aggressive, but as we built all the packages in `.cache/pikaur/pkg` ourselves, we should be able to rebuilt them again if needed),
- Keep 1 version of all packages (installed or not),
- Keep 3 versions of all packages (current default of paccache),
- or any other combo…

**Comment by KopfKrieg** [comment]
> What would be the best default parameters for automatic weekly cleanup?

Well, personally I want to keep the current and the previous version, this is also what pacman-cleanup-hook does. Maybe make that the default (or maybe make three versions the default, the current included, like paccache does).

For uninstalled packages I'm not sure. I'd remove them by default, but I think an option/switch would be better for it. Or just ask the user :)

**Comment by audeoudh** [comment]
Hard to provide an option/switch in a systemd service.

I'll use a non-aggressive as a default: 3 for installed packages (paccache's default), 1 for uninstalled packages.  If the user wants a more aggressive behaviour, it should edit the systemd unit (using `systemctl --user edit --full pikaur-cache.service` for example).

**Comment by KopfKrieg** [comment]
> Hard to provide an option/switch in a systemd service.

Not really, systemd units are able to use variables ;)

General question regarding the timer: Why do we need a *timer*? Wouldn't it be better to run a hook? E.g. pikaur could trigger a cleanup after a install/upgrade/remove operation.

**Comment by audeoudh** [comment]
Timers and hooks are not fired at the same time.

A hook is fired exactly at the end of the maintenance process, so exactly when a change occurs, cool.  But if your maintenance process is composed in many sub-steps, it will clean the cache in-between the steps, and you may not want that.

> While debugging an application on your system, you may want to re-install it, and if that is done in two steps (for any reason, e.g. because of an annoying dependency), the uninstalled package is dropped in-between.
> 
> I personally fell into the trap once.

A timer is fired independently of the maintenance process (this is currently not enforced, but probability is small that they happen in the same time), so that problem should not occur.  It's a weekly timer, so it shouldn't be too resource-consuming.

I personally prefer the timer, but it's a personal point of view.  Pikaur does provide a hook also, cf. the man page, at "How to clean old or uninstalled AUR packages in ~/.cache/pikaur/pkg?" (I updated this section in the PR).

**Comment by audeoudh** [comment]
Hum…  Maybe we can configure the hook (similar to [pacman-cleanup-hook](https://aur.archlinux.org/packages/pacman-cleanup-hook/)) to activate the cleanup service.  Thus, we have one configuration point (the service), and 3 firing points that can be activated or not: the timer (done in my PR), the hook (this suggestion), and manually with `systemctl --user start pikaur-cache.service`.  The user will choose which one to use.

What do you think about that?

**Comment by audeoudh** [comment]
I do agree with you, it would be preferable to be event-based.  But the event should be "The maintenance process ended", and we don't have that event available, excepted by manually launching `pikaur -Scc`/`paccache`.  Pacman hooks correspond to the event "The pacman command ended", which may not be the same event.

**Comment by audeoudh** [comment]
> Hum… Maybe we can configure the hook (similar to [pacman-cleanup-hook](https://aur.archlinux.org/packages/pacman-cleanup-hook/)) to activate the cleanup service. Thus, we have one configuration point (the service), and 2 firing points that can be activated or not: the timer (done in my PR), the hook (this suggestion)—and also manually, but it does the same as `pikaur -Sc`. The user will choose which one to use.
> 
> What do you think about that?

Pacman hooks are system-wide.  It may not be the best idea to clean user's cache from a system hook.  I would prefer a hook launched _by the user that started pikaur_, instead of by root that started pacman.

So I won't implement a pacman hook, until having a better idea.

**Comment by actionless** [comment]
@audeoudh  would you keep your work in the same PR (#507) or gonna create a new one?

**Comment by audeoudh** [comment]
I'll keep it thus.  Current implementation with timers seems good; implementation based on event is not ready at all.

**Comment by actionless** [comment]
fixed in #507
